# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [1]:
// TODO: Replace <FILL IN> with appropriate code

val housing = spark.read.format("csv").option("inferSchema", "true").option("header", "true").load("data/housing.csv")

housing = [longitude: double, latitude: double ... 8 more fields]


[longitude: double, latitude: double ... 8 more fields]

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [2]:
// TODO: Replace <FILL IN> with appropriate code

housing.printSchema()

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



Print the number of records in the dataset.

In [3]:
// TODO: Replace <FILL IN> with appropriate code

housing.count()

20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [4]:
// TODO: Replace <FILL IN> with appropriate code

housing.take(5)

+---------+-------+------+--------+--------+--------+--------+--------+----------+----------+
| -122.23 | 37.88 | 41.0 | 880.0  | 129.0  | 322.0  | 126.0  | 8.3252 | 452600.0 | NEAR BAY |
| -122.22 | 37.86 | 21.0 | 7099.0 | 1106.0 | 2401.0 | 1138.0 | 8.3014 | 358500.0 | NEAR BAY |
| -122.24 | 37.85 | 52.0 | 1467.0 | 190.0  | 496.0  | 177.0  | 7.2574 | 352100.0 | NEAR BAY |
| -122.25 | 37.85 | 52.0 | 1274.0 | 235.0  | 558.0  | 219.0  | 5.6431 | 341300.0 | NEAR BAY |
| -122.25 | 37.85 | 52.0 | 1627.0 | 280.0  | 565.0  | 259.0  | 3.8462 | 342200.0 | NEAR BAY |
+---------+-------+------+--------+--------+--------+--------+--------+----------+----------+

Print the number of records with population more than 10000.

In [5]:
// TODO: Replace <FILL IN> with appropriate code

housing.filter("population > 10000").count

23

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [6]:
// TODO: Replace <FILL IN> with appropriate code

housing.describe("housing_median_age", "total_rooms", "median_house_value","population").show()

+-------+------------------+------------------+------------------+------------------+
|summary|housing_median_age|       total_rooms|median_house_value|        population|
+-------+------------------+------------------+------------------+------------------+
|  count|             20640|             20640|             20640|             20640|
|   mean|28.639486434108527|2635.7630813953488|206855.81690891474|1425.4767441860465|
| stddev| 12.58555761211163|2181.6152515827944|115395.61587441359|  1132.46212176534|
|    min|               1.0|               2.0|           14999.0|               3.0|
|    max|              52.0|           39320.0|          500001.0|           35682.0|
+-------+------------------+------------------+------------------+------------------+



Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [7]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions._

housing.select(max("housing_median_age"), min("total_rooms"), mean("median_house_value")).show()

+-----------------------+----------------+-----------------------+
|max(housing_median_age)|min(total_rooms)|avg(median_house_value)|
+-----------------------+----------------+-----------------------+
|                   52.0|             2.0|     206855.81690891474|
+-----------------------+----------------+-----------------------+



## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [8]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(count("households").alias("number_houses")).sort(desc("number_houses")).show()

+---------------+-------------+
|ocean_proximity|number_houses|
+---------------+-------------+
|      <1H OCEAN|         9136|
|         INLAND|         6551|
|     NEAR OCEAN|         2658|
|       NEAR BAY|         2290|
|         ISLAND|            5|
+---------------+-------------+



Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [9]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(mean("median_house_value").alias("avg_value")).show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



Rewrite the above question in SQL.

In [10]:
// TODO: Replace <FILL IN> with appropriate code

housing.createOrReplaceTempView("df")
spark.sql("SELECT ocean_proximity, AVG(median_house_value) AS avg_value from df  GROUP BY ocean_proximity").show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [11]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.VectorAssembler

val va = new VectorAssembler().setInputCols(Array("housing_median_age", "total_rooms", "median_house_value", "population")).setOutputCol("vector")

val housingAttrs = va.transform(housing)

housingAttrs.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|              vector|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|[41.0,880.0,45260...|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|[21.0,7099.0,3585...|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|[52.0,1467.0,3521...|
|  -122.25|   37.85|              52.0|     12

va = vecAssembler_3b50247e027e
housingAttrs = [longitude: double, latitude: double ... 9 more fields]


[longitude: double, latitude: double ... 9 more fields]

In [12]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(housingAttrs, "vector").head

println(s"The standard correlation coefficient:\n ${coeff}")

The standard correlation coefficient:
 1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293   
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772    
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235  
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0                   


coeff = 


1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293   
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772    
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235  
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0                   

## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [13]:
// TODO: Replace <FILL IN> with appropriate code

val housingCol1 = housing.withColumn("rooms_per_household", expr("total_rooms/households"))
val housingCol2 = housingCol1.withColumn("bedrooms_per_room", expr("total_bedrooms/total_rooms"))
val housingExtra = housingCol2.withColumn("population_per_household", expr("population/households"))

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingExtra = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [14]:
// TODO: Replace <FILL IN> with appropriate code

val renamedHousing = housingExtra.withColumnRenamed("median_house_value", "label")

renamedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [15]:
// label columns
val colLabel = "label"

// categorical columns
val colCat = "ocean_proximity"

// numerical columns
val colNum = renamedHousing.columns.filter(_ != colLabel).filter(_ != colCat)

colLabel = label
colCat = ocean_proximity
colNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)


[longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household]

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [16]:
// TODO: Replace <FILL IN> with appropriate code

for (c <- colNum) {
    println(c + ":" + renamedHousing.filter(renamedHousing(c).isNull || renamedHousing(c) === "").count)
}

longitude:0
latitude:0
housing_median_age:0
total_rooms:0
total_bedrooms:207
population:0
households:0
median_income:0
rooms_per_household:0
bedrooms_per_room:207
population_per_household:0


As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [17]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.Imputer

val imputer = new Imputer().setStrategy("median").setInputCols(Array("total_bedrooms", "bedrooms_per_room"))
            .setOutputCols(Array("total_bedrooms", "bedrooms_per_room"))                       
val imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows



imputer = imputer_4fe02e89755c
imputedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [18]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler().setInputCols(colNum).setOutputCol("va_features")
val featuredHousing = va.transform(imputedHousing)

val scaler = new StandardScaler().setInputCol("va_features").setOutputCol("scaled_features")
val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|         va_features|     scaled_features|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.007269596069...|
|  -122.22|   37.86|              21.0|     7099.0|        1

va = vecAssembler_e8f955ccaf2d
featuredHousing = [longitude: double, latitude: double ... 12 more fields]
scaler = stdScal_178943969320
scaledHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [19]:
// TODO: Replace <FILL IN> with appropriate code

renamedHousing.select("ocean_proximity").distinct().show()

+---------------+
|ocean_proximity|
+---------------+
|         ISLAND|
|     NEAR OCEAN|
|       NEAR BAY|
|      <1H OCEAN|
|         INLAND|
+---------------+



### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [20]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.StringIndexer

val indexer = new StringIndexer().setInputCol("ocean_proximity").setOutputCol("ocean_proximity_index")
val idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

indexer = strIdx_a37ccae44ed7
idxHousing = [longitude: double, latitude: double ... 12 more fields]


[longitude: double, latitude: double ... 12 more fields]

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [21]:
indexer.fit(renamedHousing).labels

[<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND]

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [22]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoder = new OneHotEncoderEstimator().setInputCols(Array("ocean_proximity_index")).setOutputCols(Array("ocean_proximity_index_oh"))
val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------------------+------------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_proximity_index|ocean_proximity_index_oh|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------------------+------------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|                  3.0|           (4,[3],[1.0])|
|  -122.22|   37.86|              21.0| 

encoder = oneHotEncoder_c1ee23240dda
ohHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [23]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

val numPipeline = new Pipeline().setStages(Array(imputer, va, scaler))
val catPipeline = new Pipeline().setStages(Array(indexer, encoder))
val pipeline = new Pipeline().setStages(Array(numPipeline, catPipeline))
val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+---------------------+------------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|         va_features|     scaled_features|ocean_proximity_index|ocean_proximity_index_oh|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+---------------------+------------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659

numPipeline = pipeline_3357a24ce86b
catPipeline = pipeline_2e3125f5d996
pipeline = pipeline_12b4eb4348ee
newHousing = [longitude: double, latitude: double ... 15 more fields]


[longitude: double, latitude: double ... 15 more fields]

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [24]:
// TODO: Replace <FILL IN> with appropriate code

val va2 = new VectorAssembler().setInputCols(Array("scaled_features", "ocean_proximity_index_oh")).setOutputCol("features")
val dataset = va2.transform(newHousing).select("features", "label")

dataset.show(5)

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-61.007269596069...|452600.0|
|[-61.002278409814...|358500.0|
|[-61.012260782324...|352100.0|
|[-61.017251968579...|341300.0|
|[-61.017251968579...|342200.0|
+--------------------+--------+
only showing top 5 rows



va2 = vecAssembler_2c7bb30263e8
dataset = [features: vector, label: double]


[features: vector, label: double]

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [25]:
// TODO: Replace <FILL IN> with appropriate code

val Array(trainSet, testSet) = dataset.randomSplit(Array(0.8, 0.2))

trainSet = [features: vector, label: double]
testSet = [features: vector, label: double]


[features: vector, label: double]

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [26]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.LinearRegression

// train the model
val lr = new LinearRegression()
val lrModel = lr.fit(trainSet)
val trainingSummary = lrModel.summary

println(s"Coefficients: ${lrModel.coefficients}, Intercept: ${lrModel.intercept}")
println(s"RMSE: ${trainingSummary.rootMeanSquaredError}")

Coefficients: [-56338.290535541164,-57023.441576631456,13718.985817436851,8690.807149828102,1102.990135140153,-52434.74435639281,46999.65209413442,77034.48357094015,6492.047592446167,15889.408416433376,1012.2914345367736,-174817.3279234631,-209772.6859584972,-170662.3382969223,-180152.95912584083], Intercept: -2286998.65967401
RMSE: 67865.99110204013


lr = linReg_3cfe32a5ad26
lrModel = linReg_3cfe32a5ad26
trainingSummary = org.apache.spark.ml.regression.LinearRegressionTrainingSummary@5b970f50


org.apache.spark.ml.regression.LinearRegressionTrainingSummary@5b970f50

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [27]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.evaluation.RegressionEvaluator

// make predictions on the test data
val predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 190496.7750113979| 90100.0|[-61.985542102068...|
|181503.05797542492|107000.0|[-61.980550915813...|
|148905.04574355064| 67000.0|[-61.980550915813...|
|198687.34094356047|128900.0|[-61.975559729558...|
|125846.69938730216|105900.0|[-61.975559729558...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 68177.97356474373


predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_c9f4fdfc904f
rmse = 68177.97356474373


68177.97356474373

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [28]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.DecisionTreeRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val dt = new DecisionTreeRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSet)

// make predictions on the test data
val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|225973.06894100923| 90100.0|[-61.985542102068...|
| 174098.4837310195|107000.0|[-61.980550915813...|
|144547.51104100945| 67000.0|[-61.980550915813...|
|225973.06894100923|128900.0|[-61.975559729558...|
|138048.66666666666|105900.0|[-61.975559729558...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 66130.47617039186


dt = dtr_1f6ce4d3ab28
dtModel = DecisionTreeRegressionModel (uid=dtr_1f6ce4d3ab28) of depth 5 with 63 nodes
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_8bb016d6169c
rmse = 66130.47617039186


66130.47617039186

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [29]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.RandomForestRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val rf = new RandomForestRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val rfModel = rf.fit(trainSet)

// make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 194926.9161748629| 90100.0|[-61.985542102068...|
|163726.89443316354|107000.0|[-61.980550915813...|
|150528.27488519571| 67000.0|[-61.980550915813...|
|211557.44217441976|128900.0|[-61.975559729558...|
|153619.57828024338|105900.0|[-61.975559729558...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 64417.244151134684


rf = rfr_27a880e9807d
rfModel = RandomForestRegressionModel (uid=rfr_27a880e9807d) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_3bd61e383e56
rmse = 64417.244151134684


64417.244151134684

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [30]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.GBTRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val gb = new GBTRegressor().setLabelCol("label").setFeaturesCol("features")

// train the model
val gbModel = gb.fit(trainSet)

// make predictions on the test data
val predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|111666.92910490364| 90100.0|[-61.985542102068...|
|119015.53943776409|107000.0|[-61.980550915813...|
| 85230.28566984947| 67000.0|[-61.980550915813...|
|117191.79602367093|128900.0|[-61.975559729558...|
| 95148.76508960249|105900.0|[-61.975559729558...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 55677.24660201789


gb = gbtr_4b95dcca7244
gbModel = GBTRegressionModel (uid=gbtr_4b95dcca7244) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_b7061fe4a4c3
rmse = 55677.24660201789


55677.24660201789

---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [31]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder()
.addGrid(rf.numTrees, Array(1, 5, 10))
.addGrid(rf.maxDepth, Array(5, 10, 15))
.build()

val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val cv = new CrossValidator().setEstimator(rf).setEstimatorParamMaps(paramGrid).setEvaluator(evaluator).setNumFolds(3)
val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 153778.3333333333| 90100.0|[-61.985542102068...|
|100904.56140350876|107000.0|[-61.980550915813...|
| 77192.29853479854| 67000.0|[-61.980550915813...|
|183647.91919191918|128900.0|[-61.975559729558...|
|  96508.8564102564|105900.0|[-61.975559729558...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 52733.029293291846


paramGrid = 


Array({
	rfr_27a880e9807d-maxDepth: 5,
	rfr_27a880e9807d-numTrees: 1
}, {
	rfr_27a880e9807d-maxDepth: 5,
	rfr_27a880e9807d-numTrees: 5
}, {
	rfr_27a880e9807d-maxDepth: 5,
	rfr_27a880e9807d-numTrees: 10
}, {
	rfr_27a880e9807d-maxDepth: 10,
	rfr_27a880e9807d-numTrees: 1
}, {
	rfr_27a880e9807d-maxDepth: 10,
	rfr_27a880e9807d-numTrees: 5
}, {
	rfr_27a880e9807d-maxDepth: 10,
	rfr_27a880e9807d-numTrees: 10
}, {
	rfr_27a880e9807d-maxDepth: 15,
	rfr_27a880e9807d-numTrees: 1
}, {
	rfr_27a880e9807d-maxDepth: 15,
	rfr_27a880e9807d-numTrees: 5
}, {
	rfr_27a880e9807d-maxDepth: 15,
	rfr_27a8...


[{
	rfr_27a880e9807d-maxDepth: 5,
	rfr_27a880e9807d-numTrees: 1
}, {
	rfr_27a880e9807d-maxDepth: 5,
	rfr_27a880e9807d-numTrees: 5
}, {
	rfr_27a880e9807d-maxDepth: 5,
	rfr_27a880e9807d-numTrees: 10
}, {
	rfr_27a880e9807d-maxDepth: 10,
	rfr_27a880e9807d-numTrees: 1
}, {
	rfr_27a880e9807d-maxDepth: 10,
	rfr_27a880e9807d-numTrees: 5
}, {
	rfr_27a880e9807d-maxDepth: 10,
	rfr_27a880e9807d-numTrees: 10
}, {
	rfr_27a880e9807d-maxDepth: 15,
	rfr_27a880e9807d-numTrees: 1
}, {
	rfr_27a880e9807d-maxDepth: 15,
	rfr_27a880e9807d-numTrees: 5
}, {
	rfr_27a880e9807d-maxDepth: 15,
	rfr_27a880e9807d-numTrees: 10
}]

---
# 7. Custom transformer
At the end of part two, we added extra columns to the `housing` dataset. Here, we are going to implement a Transformer to do the same task. The Transformer should take the name of two input columns `inputCol1` and `inputCol2`, as well as the name of ouput column `outputCol`. It, then, computes `inputCol1` divided by `inputCol2`, and adds its result as a new column to the dataset. The details of the implemeting a custom Tranfomer is explained [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types). Please read it before before starting to implement it.

First, define the given parameters of the Transformer and implement a method to validate their schemas (`StructType`).

In [32]:
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}

trait MyParams extends Params {
    final val inputCol1 = new Param[String](this, "inputCol1", "input_column1")
    final val inputCol2 = new Param[String](this, "inputCol2", "input_column2")
    final val outputCol = new Param[String](this, "inputCol3", "input_column3")
    
  protected def validateAndTransformSchema(schema: StructType): StructType = {
    val ip1 = schema.fieldIndex($(inputCol1))
    val val1 = schema.fields(ip1)
      
    val ip2 = schema.fieldIndex($(inputCol2))
    val val2 = schema.fields(ip2)  
      
    if (val1.dataType != DoubleType) {
      throw new Exception(s"Type mismatch for ${val1.dataType} w.r.t expected input type:Double")
    }
    
    if (val2.dataType != DoubleType) {
      throw new Exception(s"Type mismatch for ${val2.dataType} w.r.t expected input type:Double")
    }
      
    schema.add(StructField($(outputCol), DoubleType, false))
  }
}

defined trait MyParams


Then, extend the class `Transformer`, and implement its setter functions for the input and output columns, and call then `setInputCol1`, `setInputCol2`, and `setOutputCol`. Morever, you need to override the methods `copy`, `transformSchema`, and the `transform`. The details of what you need to cover in these methods is given [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types).

In [33]:
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.param.{ParamMap, Param, Params}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types.{StructType, StructField, DoubleType}
import org.apache.spark.sql.functions.{col, udf}

class MyTransformer(override val uid: String) extends Transformer with MyParams {
    def this() = this(Identifiable.randomUID("configurablewordcount"))
    
    def setInputCol1(value: String): MyTransformer = {set(inputCol1, value)}
    
    def setInputCol2(value: String): MyTransformer = {set(inputCol2, value)}
    
    def setOutputCol(value: String): MyTransformer = {set(outputCol, value)}

    override def copy(extra: ParamMap): MyTransformer = {defaultCopy(extra)}

    override def transformSchema(schema: StructType) = {
        val ip1 = schema.fieldIndex($(inputCol1))
        val val1 = schema.fields(ip1)
      
        val ip2 = schema.fieldIndex($(inputCol2))
        val val2 = schema.fields(ip2)  
      
        if (val1.dataType != DoubleType) {
          throw new Exception(s"Type mismatch for ${val1.dataType} w.r.t expected input type:Double")
        }
    
        if (val2.dataType != DoubleType) {
          throw new Exception(s"Type mismatch for ${val2.dataType} w.r.t expected input type:Double")
        }
      
        schema.add(StructField($(outputCol), DoubleType, false))
    }
    
    override def transform(dataset: Dataset[_]): DataFrame = {
        val divider = udf{(column1: Double, column2: Double) => column1/column2}
        dataset.select(col("*"), divider(dataset($(inputCol1)),dataset($(inputCol2))).as($(outputCol)))
  }
}

defined class MyTransformer


Now, an instance of `MyTransformer`, and set the input columns `total_rooms` and `households`, and the output column `rooms_per_household` and run it over the `housing` dataset.

In [34]:
val myTransformer = new MyTransformer().setInputCol1("total_rooms").setInputCol2("households").setOutputCol("rooms_per_household")

val myDataset = myTransformer.transform(housing).select("rooms_per_household").show(5)

+-------------------+
|rooms_per_household|
+-------------------+
|  6.984126984126984|
|  6.238137082601054|
|  8.288135593220339|
| 5.8173515981735155|
|  6.281853281853282|
+-------------------+
only showing top 5 rows



myTransformer = configurablewordcount_2122f6dfe951


myDataset: Unit = ()


configurablewordcount_2122f6dfe951

---
# 8. Custom estimator (predictor)
Now, it's time to implement your own linear regression with gradient descent algorithm as a brand new Estimator. The whole code of the Estimator is given to you, and you do not need to implement anything. It is just a sample that shows how to build a custom Estimator.

The gradient descent update for linear regression is:
$$
w_{i+1} = w_{i} - \alpha_{i} \sum\limits_{j=1}^n (w_i^\top x_j - y_j)x_j
$$

where $i$ is the iteration number of the gradient descent algorithm, and $j$ identifies the observation. Here, $w$ represents an array of weights that is the same size as the array of features and provides a weight for each of the features when finally computing the label prediction in the form:

$$
prediction = w^\top \cdot\ x
$$

where $w$ is the final array of weights computed by the gradient descent, $x$ is the array of features of the observation point and $prediction$ is the label we predict should be associated to this observation.

The given `Helper` class implements the helper methods:
* `dot`: implements the dot product of two vectors and the dot product of a vector and a scalar
* `sum`: implements addition of two vectors
* `fill`: creates a vector of predefined size and initialize it with the predefined value

What you need to do is to implement the methods of the Linear Regresstion class `LR`, which are
* `rmsd`: computes the Root Mean Square Error of a given RDD of tuples of (label, prediction) using the formula:
$$
rmse = \sqrt{\frac{\sum\limits_{i=1}^n (label - prediction)^2}{n}}
$$
* `gradientSummand`: computes the following formula:
$$
gs_{ij} = (w_i^\top x_j - y_j)x_j
$$
* `gradient`: computes the following formula:
$$
gradient = \sum\limits_{j=1}^n gs_{ij}
$$

In [ ]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

case class Instance(label: Double, features: Vector)

object Helper extends Serializable {
  def dot(v1: Vector, v2: Vector): Double = {
    val m = Matrices.dense(1, v1.size, v1.toArray)
    m.multiply(v2).values(0)
  }

  def dot(v: Vector, s: Double): Vector = {
    val baseArray = v.toArray.map(vi => vi * s)
    Vectors.dense(baseArray)
  }

  def sumVectors(v1: Vector, v2: Vector): Vector = {
    val baseArray = ((v1.toArray) zip (v2.toArray)).map { case (val1, val2) => val1 + val2 }
    Vectors.dense(baseArray)
  }

  def fillVector(size: Int, fillVal: Double): Vector = Vectors.dense(Array.fill[Double](size)(fillVal));
}

In [ ]:
class LR() extends Serializable {
  def calcRMSE(labelsAndPreds: RDD[(Double, Double)]): Double = {
    val regressionMetrics = new RegressionMetrics(labelsAndPreds)
    regressionMetrics.rootMeanSquaredError
  }
  
  def gradientSummand(weights: Vector, lp: Instance): Vector = {
    val mult = (Helper.dot(weights, lp.features) - lp.label)
    val seq = (0 to lp.features.size - 1).map(i => lp.features(i) * mult)
    return Vectors.dense(seq.toArray)
  }
  
  def linregGradientDescent(trainData: RDD[Instance], numIters: Int): (Vector, Array[Double]) = {
    val n = trainData.count()
    val d = trainData.take(1)(0).features.size
    var w = Helper.fillVector(d, 0)
    val alpha = 1.0
    val errorTrain = Array.fill[Double](numIters)(0.0)

    for (i <- 0 until numIters) {
      val labelsAndPredsTrain = trainData.map(lp => (lp.label, Helper.dot(w, lp.features)))
      errorTrain(i) = calcRMSE(labelsAndPredsTrain)

      val gradient = trainData.map(lp => gradientSummand(w, lp)).reduce((v1, v2) => Helper.sumVectors(v1, v2))
      val alpha_i = alpha / (n * scala.math.sqrt(i + 1))
      val wAux = Helper.dot(gradient, (-1) * alpha_i)
      w = Helper.sumVectors(w, wAux)
    }
    (w, errorTrain)
  }
}

In [ ]:
abstract class MyLinearModel[FeaturesType, Model <: MyLinearModel[FeaturesType, Model]]
  extends PredictionModel[FeaturesType, Model] {
}

class MyLinearModelImpl(override val uid: String, val weights: Vector, val trainingError: Array[Double])
    extends MyLinearModel[Vector, MyLinearModelImpl] {

  override def copy(extra: ParamMap): MyLinearModelImpl = defaultCopy(extra)

  def predict(features: Vector): Double = {
    println("Predicting")
    val prediction = Helper.dot(weights, features)
    prediction
  }
}

In [ ]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._

import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearRegression[
    FeaturesType,
    Learner <: MyLinearRegression[FeaturesType, Learner, Model],
    Model <: MyLinearModel[FeaturesType, Model]]
  extends Predictor[FeaturesType, Learner, Model] {
}

class MyLinearRegressionImpl(override val uid: String)
    extends MyLinearRegression[Vector, MyLinearRegressionImpl, MyLinearModelImpl] {
  def this() = this(Identifiable.randomUID("linReg"))

  override def copy(extra: ParamMap): MyLinearRegressionImpl = defaultCopy(extra)
  
  def train(dataset: Dataset[_]): MyLinearModelImpl = {
    println("Training")

    val numIters = 10

    val instances: RDD[Instance] = dataset.select(
      col($(labelCol)), col($(featuresCol))).rdd.map {
        case Row(label: Double, features: Vector) =>
          Instance(label, features)
      }

    val (weights, trainingError) = new LR().linregGradientDescent(instances, numIters)

    new MyLinearModelImpl(uid, weights, trainingError)
  }
}

In [ ]:
import org.apache.spark.ml.evaluation.RegressionEvaluator

val lr = new MyLinearRegressionImpl().setLabelCol("label").setFeaturesCol("features")
val model = lr.fit(trainSet)
val predictions = model.transform(trainSet)
predictions.select("prediction", "label", "features").show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

---
# 9. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

## Load the data

In [35]:
val ccdefault = spark.read.format("csv").option("inferSchema", "true").option("header", "true").load("data/ccdefault.csv")

ccdefault = [ID: int, LIMIT_BAL: int ... 23 more fields]


[ID: int, LIMIT_BAL: int ... 23 more fields]

### Attribute Information
This research employed a binary variable, default payment (Yes = 1, No = 0), as the response variable. This study reviewed the literature and used the following 23 variables as explanatory variables: <br>
X1: Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit. <br>
X2: Gender (1 = male; 2 = female). <br>
X3: Education (1 = graduate school; 2 = university; 3 = high school; 4 = others). <br>
X4: Marital status (1 = married; 2 = single; 3 = others). <br>
X5: Age (year). <br>
X6 - X11: History of past payment. We tracked the past monthly payment records (from April to September, 2005) as follows: X6 = the repayment status in September, 2005; X7 = the repayment status in August, 2005; . . .;X11 = the repayment status in April, 2005. The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above. <br>
X12-X17: Amount of bill statement (NT dollar). X12 = amount of bill statement in September, 2005; X13 = amount of bill statement in August, 2005; . . .; X17 = amount of bill statement in April, 2005. <br>
X18-X23: Amount of previous payment (NT dollar). X18 = amount paid in September, 2005; X19 = amount paid in August, 2005; . . .;X23 = amount paid in April, 2005. 

## Exploratory analyses

In [36]:
ccdefault.printSchema()

root
 |-- ID: integer (nullable = true)
 |-- LIMIT_BAL: integer (nullable = true)
 |-- SEX: integer (nullable = true)
 |-- EDUCATION: integer (nullable = true)
 |-- MARRIAGE: integer (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- PAY_0: integer (nullable = true)
 |-- PAY_2: integer (nullable = true)
 |-- PAY_3: integer (nullable = true)
 |-- PAY_4: integer (nullable = true)
 |-- PAY_5: integer (nullable = true)
 |-- PAY_6: integer (nullable = true)
 |-- BILL_AMT1: integer (nullable = true)
 |-- BILL_AMT2: integer (nullable = true)
 |-- BILL_AMT3: integer (nullable = true)
 |-- BILL_AMT4: integer (nullable = true)
 |-- BILL_AMT5: integer (nullable = true)
 |-- BILL_AMT6: integer (nullable = true)
 |-- PAY_AMT1: integer (nullable = true)
 |-- PAY_AMT2: integer (nullable = true)
 |-- PAY_AMT3: integer (nullable = true)
 |-- PAY_AMT4: integer (nullable = true)
 |-- PAY_AMT5: integer (nullable = true)
 |-- PAY_AMT6: integer (nullable = true)
 |-- DEFAULT: integer (nullable = tru

In [37]:
// number of records
ccdefault.count()

30000

In [38]:
// print 5 records
ccdefault.take(5)

+-----+--------+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-------+-------+-------+-------+-------+-------+------+-------+-------+------+------+------+-----+
| 1   | 20000  | 2   | 2   | 1   | 24  | 2   | 2   | -1  | -1  | -2  | -2  | 3913  | 3102  | 689   | 0     | 0     | 0     | 0    | 689   | 0     | 0    | 0    | 0    | 1   |
| 2   | 120000 | 2   | 2   | 2   | 26  | -1  | 2   | 0   | 0   | 0   | 2   | 2682  | 1725  | 2682  | 3272  | 3455  | 3261  | 0    | 1000  | 1000  | 1000 | 0    | 2000 | 1   |
| 3   | 90000  | 2   | 2   | 2   | 34  | 0   | 0   | 0   | 0   | 0   | 0   | 29239 | 14027 | 13559 | 14331 | 14948 | 15549 | 1518 | 1500  | 1000  | 1000 | 1000 | 5000 | 0   |
| 4   | 50000  | 2   | 2   | 1   | 37  | 0   | 0   | 0   | 0   | 0   | 0   | 46990 | 48233 | 49291 | 28314 | 28959 | 29547 | 2000 | 2019  | 1200  | 1100 | 1069 | 1000 | 0   |
| 5   | 50000  | 1   | 2   | 1   | 57  | -1  | 0   | -1  | 0   | 0   | 0   | 8617  | 5670  | 35835 | 20940 | 19146 | 19131 | 2000 | 36681 | 10000 | 9000 | 689  | 679  | 0   |
+-----+--------+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-------+-------+-------+-------+-------+-------+------+-------+-------+------+------+------+-----+

In [39]:
// summary of age attribute
ccdefault.describe("AGE").show()

+-------+-----------------+
|summary|              AGE|
+-------+-----------------+
|  count|            30000|
|   mean|          35.4855|
| stddev|9.217904068090155|
|    min|               21|
|    max|               79|
+-------+-----------------+



In [40]:
//summary of LIMIT_BAL attribute
ccdefault.describe("LIMIT_BAL").show()

+-------+------------------+
|summary|         LIMIT_BAL|
+-------+------------------+
|  count|             30000|
|   mean|167484.32266666667|
| stddev|129747.66156720246|
|    min|             10000|
|    max|           1000000|
+-------+------------------+



In [41]:
// summary of bill amount related attributes BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6
ccdefault.describe("BILL_AMT1", "BILL_AMT2", "BILL_AMT3", "BILL_AMT4", "BILL_AMT5", "BILL_AMT6").show()

+-------+-----------------+-----------------+-----------------+------------------+-----------------+----------------+
|summary|        BILL_AMT1|        BILL_AMT2|        BILL_AMT3|         BILL_AMT4|        BILL_AMT5|       BILL_AMT6|
+-------+-----------------+-----------------+-----------------+------------------+-----------------+----------------+
|  count|            30000|            30000|            30000|             30000|            30000|           30000|
|   mean|       51223.3309|49179.07516666667|       47013.1548| 43262.94896666666|40311.40096666667|      38871.7604|
| stddev|73635.86057552966|71173.76878252832|69349.38742703677|64332.856133916444|60797.15577026471|59554.1075367459|
|    min|          -165580|           -69777|          -157264|           -170000|           -81334|         -339603|
|    max|           964511|           983931|          1664089|            891586|           927171|          961664|
+-------+-----------------+-----------------+-----------

In [42]:
// summary of previous payment amount related attributes PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6
ccdefault.describe("PAY_AMT1", "PAY_AMT2", "PAY_AMT3", "PAY_AMT4", "PAY_AMT5", "PAY_AMT6").show()

+-------+-----------------+------------------+-----------------+------------------+------------------+-----------------+
|summary|         PAY_AMT1|          PAY_AMT2|         PAY_AMT3|          PAY_AMT4|          PAY_AMT5|         PAY_AMT6|
+-------+-----------------+------------------+-----------------+------------------+------------------+-----------------+
|  count|            30000|             30000|            30000|             30000|             30000|            30000|
|   mean|        5663.5805|         5921.1635|        5225.6815| 4826.076866666666| 4799.387633333334|5215.502566666667|
| stddev|16563.28035402577|23040.870402057186|17606.96146980311|15666.159744032062|15278.305679144742|17777.46577543531|
|    min|                0|                 0|                0|                 0|                 0|                0|
|    max|           873552|           1684259|           896040|            621000|            426529|           528666|
+-------+-----------------+-----

#### Break down the data by categorical attributes
Categorical attributes are: SEX, EDUCATION, MARRIAGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6

In [43]:
ccdefault.createOrReplaceTempView("df")

In [44]:
// Gender based distribution of default 1=male, 2=female
spark.sql("SELECT SEX, COUNT(DEFAULT) from df WHERE DEFAULT = 1 GROUP BY SEX").show()

+---+--------------+
|SEX|count(DEFAULT)|
+---+--------------+
|  1|          2873|
|  2|          3763|
+---+--------------+



In [45]:
// Education based distribution of default 
spark.sql("SELECT EDUCATION, COUNT(DEFAULT) from df WHERE DEFAULT = 1 GROUP BY EDUCATION").show()

+---------+--------------+
|EDUCATION|count(DEFAULT)|
+---------+--------------+
|        1|          2036|
|        6|             8|
|        3|          1237|
|        5|            18|
|        4|             7|
|        2|          3330|
+---------+--------------+



In [46]:
// Marriage based distribution of default 
spark.sql("SELECT MARRIAGE, COUNT(DEFAULT) from df WHERE DEFAULT = 1 GROUP BY MARRIAGE").show()

+--------+--------------+
|MARRIAGE|count(DEFAULT)|
+--------+--------------+
|       1|          3206|
|       3|            84|
|       2|          3341|
|       0|             5|
+--------+--------------+



#### Prepare the data for machine learning

In [47]:
// rename the label column from DEFAULT to label
val renamedCcdefault = ccdefault.withColumnRenamed("DEFAULT", "label")

renamedCcdefault = [ID: int, LIMIT_BAL: int ... 23 more fields]


[ID: int, LIMIT_BAL: int ... 23 more fields]

Removing ID column

In [48]:
val renamedCcdefaultNew = renamedCcdefault.drop("ID")

renamedCcdefaultNew = [LIMIT_BAL: int, SEX: int ... 22 more fields]


[LIMIT_BAL: int, SEX: int ... 22 more fields]

Separate the numerical attributes from the categorical attributes(SEX, EDUCATION, MARRIAGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6)

In [49]:
// label columns
val colLabelDefault = "label"

// categorical columns
val colCatDefault = Array("SEX", "EDUCATION", "MARRIAGE", "PAY_0", "PAY_2", "PAY_3", "PAY_4", "PAY_5", "PAY_6")

// numerical columns
val colNumDefault = renamedCcdefaultNew.columns.filter(_ != colLabelDefault).filter(!colCatDefault.contains(_))

colLabelDefault = label
colCatDefault = Array(SEX, EDUCATION, MARRIAGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6)
colNumDefault = Array(LIMIT_BAL, AGE, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)


[LIMIT_BAL, AGE, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6]

Checking for missing values

In [50]:
for (c <- colNumDefault) {
    println(c + ":" + renamedCcdefaultNew.filter(renamedCcdefaultNew(c).isNull || renamedCcdefaultNew(c) === "").count)
}

LIMIT_BAL:0
AGE:0
BILL_AMT1:0
BILL_AMT2:0
BILL_AMT3:0
BILL_AMT4:0
BILL_AMT5:0
BILL_AMT6:0
PAY_AMT1:0
PAY_AMT2:0
PAY_AMT3:0
PAY_AMT4:0
PAY_AMT5:0
PAY_AMT6:0


### Feature scaling

In [83]:
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val vaCCDefault = new VectorAssembler().setInputCols(colNumDefault).setOutputCol("va_features")
val featuredCCDefault = vaCCDefault.transform(renamedCcdefaultNew)

val scalerCCDefault = new StandardScaler().setInputCol("va_features").setOutputCol("scaled_features")
val scaledCCDefault = scalerCCDefault.fit(featuredCCDefault).transform(featuredCCDefault)

scaledCCDefault.show(5)

+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+--------------------+--------------------+
|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|         va_features|     scaled_features|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+--------------------+--------------------+
|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|    1|(14,[0,1,2,3,4,9]...|(14,[0,1,2,3,4,9]...|
|   120000|  2|        2|       2| 26|   -1|    2|  

vaCCDefault = vecAssembler_d950eec08f5a
featuredCCDefault = [LIMIT_BAL: int, SEX: int ... 23 more fields]
scalerCCDefault = stdScal_ae84589f48a6
scaledCCDefault = [LIMIT_BAL: int, SEX: int ... 24 more fields]


[LIMIT_BAL: int, SEX: int ... 24 more fields]

In [84]:
import org.apache.spark.ml.feature.StringIndexer

val indexerDefault0 = new StringIndexer().setInputCol("PAY_0").setOutputCol("PAY_0_index")
val indexerDefault2 = new StringIndexer().setInputCol("PAY_2").setOutputCol("PAY_2_index")
val indexerDefault3 = new StringIndexer().setInputCol("PAY_3").setOutputCol("PAY_3_index")
val indexerDefault4 = new StringIndexer().setInputCol("PAY_4").setOutputCol("PAY_4_index")
val indexerDefault5 = new StringIndexer().setInputCol("PAY_5").setOutputCol("PAY_5_index")
val indexerDefault6 = new StringIndexer().setInputCol("PAY_6").setOutputCol("PAY_6_index")

val renamedCcdefaultNewLatest0 = indexerDefault0.fit(renamedCcdefaultNew).transform(renamedCcdefaultNew)
val renamedCcdefaultNewLatest2 = indexerDefault2.fit(renamedCcdefaultNewLatest0).transform(renamedCcdefaultNewLatest0)
val renamedCcdefaultNewLatest3 = indexerDefault3.fit(renamedCcdefaultNewLatest2).transform(renamedCcdefaultNewLatest2)
val renamedCcdefaultNewLatest4 = indexerDefault4.fit(renamedCcdefaultNewLatest3).transform(renamedCcdefaultNewLatest3)
val renamedCcdefaultNewLatest5 = indexerDefault5.fit(renamedCcdefaultNewLatest4).transform(renamedCcdefaultNewLatest4)
val renamedCcdefaultNewLatest6 = indexerDefault6.fit(renamedCcdefaultNewLatest5).transform(renamedCcdefaultNewLatest5)

indexerDefault0 = strIdx_2dd91dd0a00f
indexerDefault2 = strIdx_e94b72498e97
indexerDefault3 = strIdx_faab712bdb87
indexerDefault4 = strIdx_482d36f4e9ff
indexerDefault5 = strIdx_11b35778f5f4
indexerDefault6 = strIdx_a7ffd97d0570
renamedCcdefaultNewLatest0 = [LIMIT_BAL: int, SEX: int ... 23 more fields]
renamedCcdefaultNewLatest2 = [LIMIT_BAL: int, SEX: int ... 24 more fields]


renamedCcdefaultNewLatest3: org.apache.spark.sql...


[LIMIT_BAL: int, SEX: int ... 24 more fields]

In [86]:
import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoderCCDefault = new OneHotEncoderEstimator().setInputCols(Array("SEX", "EDUCATION", "MARRIAGE", "PAY_0_index", "PAY_2_index", "PAY_3_index", "PAY_4_index", "PAY_5_index", "PAY_6_index")).setOutputCols(Array("SEX_oh", "EDUCATION_oh", "MARRIAGE_oh", "PAY_0_oh", "PAY_2_oh", "PAY_3_oh", "PAY_4_oh", "PAY_5_oh", "PAY_6_oh"))
val ohCCDefault = encoderCCDefault.fit(renamedCcdefaultNewLatest6).transform(renamedCcdefaultNewLatest6)

ohCCDefault.show(5)

+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+-----------+-----------+-----------+-----------+-----------+-----------+-------------+-------------+--------------+-------------+--------------+--------------+-------------+--------------+-------------+
|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|PAY_0_index|PAY_2_index|PAY_3_index|PAY_4_index|PAY_5_index|PAY_6_index|     PAY_6_oh|     PAY_5_oh|      PAY_2_oh|  MARRIAGE_oh|      PAY_3_oh|      PAY_0_oh| EDUCATION_oh|      PAY_4_oh|       SEX_oh|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+-----------+

encoderCCDefault = oneHotEncoder_54d351f6217e
ohCCDefault = [LIMIT_BAL: int, SEX: int ... 37 more fields]


[LIMIT_BAL: int, SEX: int ... 37 more fields]

### Pipeline

In [87]:
import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

val numPipelineCCDefault = new Pipeline().setStages(Array(vaCCDefault, scalerCCDefault))
val catPipelineCCDefault = new Pipeline().setStages(Array(indexerDefault0, indexerDefault2, indexerDefault3, indexerDefault4, indexerDefault5, indexerDefault6, encoderCCDefault))
val pipelineCCDDefault = new Pipeline().setStages(Array(numPipelineCCDefault, catPipelineCCDefault))
val newCCDefault = pipelineCCDDefault.fit(renamedCcdefaultNew).transform(renamedCcdefaultNew)

newCCDefault.show(5)

+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+--------------------+--------------------+-----------+-----------+-----------+-----------+-----------+-----------+-------------+-------------+--------------+-------------+--------------+--------------+-------------+--------------+-------------+
|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|         va_features|     scaled_features|PAY_0_index|PAY_2_index|PAY_3_index|PAY_4_index|PAY_5_index|PAY_6_index|     PAY_6_oh|     PAY_5_oh|      PAY_2_oh|  MARRIAGE_oh|      PAY_3_oh|      PAY_0_oh| EDUCATION_oh|      PAY_4_oh|       SEX_oh|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+--------

numPipelineCCDefault = pipeline_17c3a6d72dd9
catPipelineCCDefault = pipeline_5fecd0b9bd84
pipelineCCDDefault = pipeline_a5cf63072d99
newCCDefault = [LIMIT_BAL: int, SEX: int ... 39 more fields]


[LIMIT_BAL: int, SEX: int ... 39 more fields]

In [88]:
newCCDefault.columns

[LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, label, va_features, scaled_features, PAY_0_index, PAY_2_index, PAY_3_index, PAY_4_index, PAY_5_index, PAY_6_index, PAY_6_oh, PAY_5_oh, PAY_2_oh, MARRIAGE_oh, PAY_3_oh, PAY_0_oh, EDUCATION_oh, PAY_4_oh, SEX_oh]

Removing old and temporary generated columns.

In [89]:
val newCCDefaultLatest = newCCDefault.drop("LIMIT_BAL", "SEX", "EDUCATION", "MARRIAGE", "PAY_0", "PAY_2", "PAY_3", "PAY_4", "PAY_5", "PAY_6", "PAY_0_index", "PAY_2_index", "PAY_3_index", "PAY_4_index", "PAY_5_index", "PAY_6_index", "va_features")

newCCDefaultLatest = [AGE: int, BILL_AMT1: int ... 22 more fields]


[AGE: int, BILL_AMT1: int ... 22 more fields]

Put all attributes of the final dataset newHousing into a big vector, and call the new column features.

In [90]:
val va2 = new VectorAssembler().setInputCols(Array("scaled_features", "SEX_oh", "EDUCATION_oh", "MARRIAGE_oh", "PAY_0_oh", "PAY_2_oh", "PAY_3_oh", "PAY_4_oh", "PAY_5_oh", "PAY_6_oh")).setOutputCol("features")
val datasetCCDefault = va2.transform(newCCDefaultLatest).select("features", "label")

datasetCCDefault.show(5)

+--------------------+-----+
|            features|label|
+--------------------+-----+
|(83,[0,1,2,3,4,9,...|    1|
|(83,[0,1,2,3,4,5,...|    1|
|(83,[0,1,2,3,4,5,...|    0|
|(83,[0,1,2,3,4,5,...|    0|
|(83,[0,1,2,3,4,5,...|    0|
+--------------------+-----+
only showing top 5 rows



va2 = vecAssembler_22e3608b2b6d
datasetCCDefault = [features: vector, label: int]


[features: vector, label: int]

## Model creation

Train test split of data

In [125]:
val Array(trainSetDefault, testSetDefault) = datasetCCDefault.randomSplit(Array(0.8, 0.2))

trainSetDefault = [features: vector, label: int]
testSetDefault = [features: vector, label: int]


[features: vector, label: int]

### Logistic Regression

In [127]:
import org.apache.spark.ml.classification.LogisticRegression

// train the model
val lr = new LogisticRegression()
val lrModel = lr.fit(trainSetDefault)

lr = logreg_29023128803b
lrModel = logreg_29023128803b


logreg_29023128803b

In [134]:
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator

val predictions = lrModel.transform(testSetDefault)

val evaluator = new BinaryClassificationEvaluator().setLabelCol("label").setRawPredictionCol("prediction").setMetricName("areaUnderROC")

val area_under_curve = evaluator.evaluate(predictions)

val label_prediction = predictions.select("label", "prediction")
val total_count = predictions.count()
val correct_prediction_count = label_prediction.filter($"label" === $"prediction").count()
val wrong_prediction_count = label_prediction.filter(not($"label" === $"prediction")).count()
val true_positive = label_prediction.filter($"prediction" === 0.0).filter($"label" === $"prediction").count()
val false_negative = label_prediction.filter($"prediction" === 0.0).filter(not($"label" === $"prediction")).count()
val false_positive = label_prediction.filter($"prediction" === 1.0).filter(not($"label" === $"prediction")).count()
val true_negative = total_count - (true_positive + false_negative + false_positive)
val precision = (true_positive.toFloat) / (true_positive + false_positive)
val recall = (true_positive.toFloat) / (true_positive + false_negative)
val wrong_to_total_ratio = wrong_prediction_count.toDouble/total_count.toDouble
val correct_to_total_ratio = correct_prediction_count.toDouble/total_count.toDouble

predictions = [features: vector, label: int ... 3 more fields]
evaluator = binEval_ae0ca54e5d49
area_under_curve = 0.656032940973772
label_prediction = [label: int, prediction: double]
total_count = 5960
correct_prediction_count = 4916
wrong_prediction_count = 1044
true_positive = 4456
false_negative = 813
false_positive = 231
true_negative = 460
precision = 0.95071477
recall = 0.8457013
wrong_to_total_ratio = 0.17516778523489934
correct_to_total_ratio = 0.8248322147651007


0.8248322147651007

### Decision Tree

In [129]:
import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier

val dt = new DecisionTreeClassifier()
  .setLabelCol("label")
  .setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSetDefault)

dt = dtc_16d1f5c82051
dtModel = DecisionTreeClassificationModel (uid=dtc_16d1f5c82051) of depth 5 with 55 nodes


DecisionTreeClassificationModel (uid=dtc_16d1f5c82051) of depth 5 with 55 nodes

In [135]:
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator

val predictions = dtModel.transform(testSetDefault)

val evaluator = new BinaryClassificationEvaluator().setLabelCol("label").setRawPredictionCol("prediction").setMetricName("areaUnderROC")

val area_under_curve = evaluator.evaluate(predictions)

val label_prediction = predictions.select("label", "prediction")
val total_count = predictions.count()
val correct_prediction_count = label_prediction.filter($"label" === $"prediction").count()
val wrong_prediction_count = label_prediction.filter(not($"label" === $"prediction")).count()
val true_positive = label_prediction.filter($"prediction" === 0.0).filter($"label" === $"prediction").count()
val false_negative = label_prediction.filter($"prediction" === 0.0).filter(not($"label" === $"prediction")).count()
val false_positive = label_prediction.filter($"prediction" === 1.0).filter(not($"label" === $"prediction")).count()
val true_negative = total_count - (true_positive + false_negative + false_positive)
val precision = (true_positive.toFloat) / (true_positive + false_positive)
val recall = (true_positive.toFloat) / (true_positive + false_negative)
val wrong_to_total_ratio = wrong_prediction_count.toDouble/total_count.toDouble
val correct_to_total_ratio = correct_prediction_count.toDouble/total_count.toDouble

predictions = [features: vector, label: int ... 3 more fields]
evaluator = binEval_84a40a816cbd
area_under_curve = 0.6541418149279206
label_prediction = [label: int, prediction: double]
total_count = 5960
correct_prediction_count = 4909
wrong_prediction_count = 1051
true_positive = 4453
false_negative = 817
false_positive = 234
true_negative = 456
precision = 0.9500747
recall = 0.84497154
wrong_to_total_ratio = 0.17634228187919462
correct_to_total_ratio = 0.8236577181208053


0.8236577181208053

### Random forest classifier

In [131]:
import org.apache.spark.ml.classification.{RandomForestClassificationModel, RandomForestClassifier}

val rf = new RandomForestClassifier()
    .setLabelCol("label")
    .setFeaturesCol("features")

// train the model
val rfModel = rf.fit(trainSetDefault)

rf = rfc_e78ebb4c3bd8
rfModel = RandomForestClassificationModel (uid=rfc_e78ebb4c3bd8) with 20 trees


RandomForestClassificationModel (uid=rfc_e78ebb4c3bd8) with 20 trees

In [136]:
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator

val predictions = rfModel.transform(testSetDefault)

val evaluator = new BinaryClassificationEvaluator().setLabelCol("label").setRawPredictionCol("prediction").setMetricName("areaUnderROC")

val area_under_curve = evaluator.evaluate(predictions)

val label_prediction = predictions.select("label", "prediction")
val total_count = predictions.count()
val correct_prediction_count = label_prediction.filter($"label" === $"prediction").count()
val wrong_prediction_count = label_prediction.filter(not($"label" === $"prediction")).count()
val true_positive = label_prediction.filter($"prediction" === 0.0).filter($"label" === $"prediction").count()
val false_negative = label_prediction.filter($"prediction" === 0.0).filter(not($"label" === $"prediction")).count()
val false_positive = label_prediction.filter($"prediction" === 1.0).filter(not($"label" === $"prediction")).count()
val true_negative = total_count - (true_positive + false_negative + false_positive)
val precision = (true_positive.toFloat) / (true_positive + false_positive)
val recall = (true_positive.toFloat) / (true_positive + false_negative)
val wrong_to_total_ratio = wrong_prediction_count.toDouble/total_count.toDouble
val correct_to_total_ratio = correct_prediction_count.toDouble/total_count.toDouble

predictions = [features: vector, label: int ... 3 more fields]
evaluator = binEval_c75553925b9e
area_under_curve = 0.6205475324018851
label_prediction = [label: int, prediction: double]
total_count = 5960
correct_prediction_count = 4873
wrong_prediction_count = 1087
true_positive = 4521
false_negative = 921
false_positive = 166
true_negative = 352
precision = 0.96458286
recall = 0.8307608
wrong_to_total_ratio = 0.18238255033557046
correct_to_total_ratio = 0.8176174496644295


0.8176174496644295

### Conclusion

Comparison of performance of different algorithms employed is given below.

<img src="performance_measure.png"></img>

As given in table above, given dataset is pre-processed to generate models using three different algorithms. We have calculated different parameters to decide best possible solution for this dataset. For Logistic regression and decision tree AUC is quite similar i.e. 0.65 and for RandomForest it is 0.62. For RandomForest, all parameters are quite low compare other two models. So Random forest is not best possible choice. For other two models, difference is not that significant, but based on number of correct predictions, Logistic Regression model will be best choice. We can do hyperparameter optimization for better outcome with data. Some more data pre-processing techniques can be experimented to achieve better results.